Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -15,176 +15,121 @@
 //******************************************************************************
 
 #import "CGContextInternal.h"
-#import "CGPathInternal.h"
-#import "CGPathInternal.h"
-#include "LoggingNative.h"
 #import <CoreGraphics/CGBitmapContext.h>
-#import <CoreGraphics/CGColorSpace.h>
-#import <CoreGraphics/CGContext.h>
-#import <CoreGraphics/CGContext.h>
-#import <CoreGraphics/CGGeometry.h>
-#import <CoreGraphics/CGGeometry.h>
-#import <CoreGraphics/CGImage.h>
-#import <CoreGraphics/CoreGraphicsExport.h>
+#import <CoreGraphics/CoreGraphics.h>
 #import <Starboard.h>
 #import <StubReturn.h>
 #import <algorithm>
 #import <vector>
 
-static const wchar_t* TAG = L"CGPath";
+#import <CoreFoundation/CoreFoundation.h>
+#import <CFRuntime.h>
+#import "D2DWrapper.h"
 
-class BBox {
-public:
-    float x, y, x1, y1;
-    bool xset, yset, x1set, y1set;
+#include <COMIncludes.h>
+#import <wrl/client.h>
+#import <D2d1.h>
+#include <COMIncludes_End.h>
+#include "CFCPPBase.h"
 
-    BBox() {
-        x = y = x1 = y1 = 0;
-        xset = yset = x1set = y1set = false;
-    }
+static const wchar_t* TAG = L"CGPath";
 
-    void addPoint(float ptX, float ptY) {
-        if (ptX < x || !xset) {
-            x = ptX;
-            xset = true;
-        }
-        if (ptY < y || !yset) {
-            y = ptY;
-            yset = true;
-        }
+using namespace std;
+using namespace Microsoft::WRL;
 
-        if (ptX > x1 || !x1set) {
-            x1 = ptX;
-            x1set = true;
-        }
+struct __CGPathImpl {
+    ComPtr<ID2D1PathGeometry> pathGeometry{ nullptr };
+    ComPtr<ID2D1GeometrySink> geometrySink{ nullptr };
 
-        if (ptY > y1 || !y1set) {
-            y1 = ptY;
-            y1set = true;
-        }
+    bool isFigureClosed;
+    CGPoint currentPoint{ 0, 0 };
+    CGPoint startingPoint{ 0, 0 };
+
+    __CGPathImpl() : isFigureClosed(true) {
     }
 };
 
-__CGPath::~__CGPath() {
-    if (_elements) {
-        IwFree(_elements);
+struct __CGPath : CoreFoundation::CppBase<__CGPath, __CGPathImpl> {
+    // A private helper function for re-opening a path geometry. CGPath does not
+    // have a concept of an open and a closed path but D2D relies on it. A
+    // path/sink cannot be read from while the path is open thus it must be
+    // closed. However, a CGPath can be edited again after being read from so
+    // we must open the path again. This cannot be done normally, so we must
+    // create a new path with the old path information to edit.
+    void preparePathForEditing() {
+        if (!_impl.geometrySink) {
+            // Re-open this geometry.
+            ComPtr<ID2D1Factory> factory = _GetD2DFactoryInstance();
+
+            // Create temp vars for new path/sink
+            Microsoft::WRL::ComPtr<ID2D1PathGeometry> newPath;
+            Microsoft::WRL::ComPtr<ID2D1GeometrySink> newSink;
```

Comment:
Sorry, I had this fix locally, forgot to push last night.

---
In reply to: [83929529](https://github.com/Microsoft/WinObjC/pull/1138#discussion_r83929529) [](ancestors = 83929529)